pkg/chain/chain.go

```diff
@@ -1,27 +1,83 @@
-// Package eth contains interface for interaction with an ethereum blockchain
-// along with structures reflecting events emitted on an ethereum blockchain.
-package eth // TODO: rename; this can be any host chain
+// Package chain contains interface for interaction with a host chain along with
+// structures reflecting events emitted on such a chain.
+package chain // TODO: rename; this can be any host chain
 
 import (
+	cecdsa "crypto/ecdsa"
+	"fmt"
 	"math/big"
 	"time"
 
-	"github.com/ethereum/go-ethereum/common"
 	"github.com/keep-network/keep-common/pkg/subscription"
 	"github.com/keep-network/keep-core/pkg/chain"
 	"github.com/keep-network/keep-ecdsa/pkg/ecdsa"
 )
 
-// Handle represents a handle to an ethereum blockchain.
+type KeepApplicationID interface {
+	fmt.Stringer
+}
+
+type KeepMemberID interface {
+	fmt.Stringer
+
+	// KeepMemberIDs should be convertible to their associated OperatorID. This
+	// may be a 1-to-1 correspondence, i.e. a simple interface cast, or it may
+	// be an internally-tracked association on the host chain.
+	OperatorID() OperatorID
+}
+
+type OperatorID interface {
+	fmt.Stringer
+
+	// OperatorIDs should be convertible to an associated KeepMemberID given the
+	// id of the keep in question. This may be a 1-to-1 correspondence, i.e. a
+	// simple interface cast, or it may be an internally-tracked association on
+	// the host chain. The conversion _may but is not guaranteed to be_ nil if
+	// the operator is not in the specified keep.
+	KeepMemberID(keepID KeepID) KeepMemberID
```

Comment:
I do not understand why do we need to pass `keepID` as a parameter. In all usages of `KeepMemberID` I looked at, this function assumes conversion succeeded.

Reply:
Yeah, it doesn't matter for Ethereum, but my observation here is that there's no reason to assume that a Keep member's id is the same as their operator id (indeed, the split in types is meant to reflect that). If it isn't guaranteed they're the same, then the Keep member's id is likely specific to the keep, in which case if we're going from operator id to keep member id, then the function would need some way to connect the operator id to the keep member id for a particular keep---necessitating a keep ID.

This might be a little bit of premature design, but it feels much more correct in the new typing structure.

Reply:
Yeah, totally makes sense. I am concerned about the potential `nil` conversion though because it means we probably need to do a post-conversion check for every call and report an error if something goes wrong. I also don't fully understand in what scenarios `nil` could be returned.

For example, if `KeepMemberID` is a mix of `KeepID` and `MemberID`, this function should always succeed. 

On the other hand, if the result depends on the on-chain state and this function is expected to validate if the given `MemberID` is really a member of a keep with the given `KeepID` before doing the conversion, we should let this function report `error`. But it feels wrong to execute such checks for conversion functions.

Reply:
The `KeepMemberID` *could be nil* if the operator is not in the keep. Right now that won't be the case in any implementations. We could deal with this using an `error`, but that felt like too much with the current setup (though it may make sense to add it later). There are very few places where this conversion happens at the moment, and I think discouraging the conversion is the right thing---this is why `IsThisOperatorMember` exists on keeps now, for example.

Frankly, I would rather _not be able to convert between these types_ -- it just wasn't feasible the way things are currently wired up.